### PR TITLE
Energy Weapon Recharging: The Future Is Now

### PR DESCRIPTION
**Pull request by RustingWithYou**
Adds two new methods for recharging energy weapons - a hardsuit module which will drain the hardsuit's power cell to charge a weapon and a recharger backpack which requires power cells.

These can be made by science/machinist with sufficient tech and materials, and are also available in the antag uplink. ERT suits, as well as relevant hardsuits for factions which primarily use energy weapons (TCFL, Skrell, Vaurca and Hegemony), have had the recharge modules added to them.

### CODE REVIEW

**Comment by RustingWithYou** [comment]
!review

File: code/game/objects/items/recharger_backpack.dm [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	connected = newgun
+
+/obj/item/recharger_backpack/verb/disconnect()
+	set name = "Disconnect Energy Weapon"
+	set category = "Object"
+	var/mob/living/carbon/human/user
+	if(istype(usr,/mob/living/carbon/human))
+		user = usr
+	else
+		return
```
Comment (FluffyGhoster) [inline code/game/objects/items/recharger_backpack.dm:56]:
```suggestion
	//If there caller isn't an human, early return
	if(!ishuman(usr))
		return

	user = usr
```

File: code/game/objects/items/recharger_backpack.dm [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+	if(!user)
+		return
+	if(!connected)
+		to_chat(usr, SPAN_WARNING("\The [src] has no energy weapon connected!"))
+		return
+	connected.disconnect()
```
Comment (FluffyGhoster) [inline code/game/objects/items/recharger_backpack.dm:62]:
```suggestion

	connected.disconnect()
```

File: code/game/objects/items/recharger_backpack.dm [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+				icon_state = "recharger_backpack"
+				item_state = "recharger_backpack"
+				set_light(0)
+
+/obj/item/recharger_backpack/Destroy()
+	. = ..()
+	if(connected)
+		connected.disconnect()
+	if(powersupply)
+		QDEL_NULL(powersupply)
```
Comment (FluffyGhoster) [inline code/game/objects/items/recharger_backpack.dm:95]:
```suggestion
	if(connected)
		connected.disconnect()

	if(powersupply)
		QDEL_NULL(powersupply)
	
	. = ..()
```

File: code/modules/clothing/spacesuits/rig/modules/utility.dm [hunk 4]
```diff
@@ -878,3 +878,32 @@ var/global/list/lattice_users = list()
...
+	desc = "A specialised power cable designed to connect an energy weapon to a hardsuit's power supply."
+	toggleable = TRUE
+	icon_state = "powersink"
+	interface_name = "integrated weapon recharger"
+	interface_desc = "Can connect to an energy weapon, recharging it off the hardsuit's power supply. Drag the weapon onto the hardsuit control module to connect it."
+	category = MODULE_LIGHT_COMBAT
+	usable = FALSE
+	disruptive = FALSE
+	confined_use = TRUE
+	var/obj/item/gun/energy/connected //the gun charging off our hardsuit
```
Comment (FluffyGhoster) [inline code/modules/clothing/spacesuits/rig/modules/utility.dm:893]:
```suggestion

	///The gun charging off our hardsuit
	var/obj/item/gun/energy/connected
```

File: code/modules/clothing/spacesuits/rig/modules/utility.dm [hunk 5]
```diff
@@ -878,3 +878,32 @@ var/global/list/lattice_users = list()
...
+
+/obj/item/rig_module/recharger/activate(mob/user)
+	if (!..())
+		return FALSE
+	if(!connected)
```
Comment (FluffyGhoster) [inline code/modules/clothing/spacesuits/rig/modules/utility.dm:898]:
```suggestion

	if(!connected)
```

File: code/modules/clothing/spacesuits/rig/modules/utility.dm [hunk 6]
```diff
@@ -878,3 +878,32 @@ var/global/list/lattice_users = list()
...
+		to_chat(user, SPAN_NOTICE("\The [src] does not have a connected energy weapon to charge!"))
+		return FALSE
+	to_chat(user, SPAN_NOTICE("\The [connected] is now connected to your hardsuit power supply. Deactivate this module to disconnect it."))
```
Comment (FluffyGhoster) [inline code/modules/clothing/spacesuits/rig/modules/utility.dm:901]:
```suggestion

	to_chat(user, SPAN_NOTICE("\The [connected] is now connected to your hardsuit power supply. Deactivate this module to disconnect it."))
```

File: code/modules/clothing/spacesuits/rig/modules/utility.dm [hunk 7]
```diff
@@ -878,3 +878,32 @@ var/global/list/lattice_users = list()
...
+	return TRUE
+
+/obj/item/rig_module/recharger/deactivate(mob/user)
+	if (!..())
+		return FALSE
+	if(connected)
```
Comment (FluffyGhoster) [inline code/modules/clothing/spacesuits/rig/modules/utility.dm:907]:
```suggestion

	if(connected)
```

File: code/modules/clothing/spacesuits/rig/modules/utility.dm [hunk 8]
```diff
@@ -878,3 +878,32 @@ var/global/list/lattice_users = list()
...
+		connected.disconnect()
+	return TRUE
```
Comment (FluffyGhoster) [inline code/modules/clothing/spacesuits/rig/modules/utility.dm:909]:
```suggestion

	return TRUE
```

File: code/modules/projectiles/guns/energy.dm [hunk 9]
```diff
@@ -132,6 +142,71 @@
...
+		if(!rigcharge.holder || !rigcharge.holder.wearer)
+			to_chat(usr, SPAN_WARNING("\The [rigcharge] must be installed and actives!"))
+			return
+		to_chat(usr, SPAN_NOTICE("You neatly plug \the [src] into \the [recharger]."))
+		playsound(get_turf(src), 'sound/machines/click.ogg', 30, 0)
+		rigcharge.connected = src
+		rigcharger = rigcharge
+		self_recharge = TRUE
+		use_external_power = TRUE
+	if(istype(recharger, /obj/item/recharger_backpack))
```
Comment (FluffyGhoster) [inline code/modules/projectiles/guns/energy.dm:160]:
```suggestion

	else if(istype(recharger, /obj/item/recharger_backpack))
```

File: code/modules/projectiles/guns/energy.dm [hunk 10]
```diff
@@ -132,6 +142,71 @@
...
+		var/obj/item/recharger_backpack/back_charge = recharger
+		if(rigcharger)
```
Comment (FluffyGhoster) [inline code/modules/projectiles/guns/energy.dm:162]:
```suggestion

		if(rigcharger)
```

File: code/modules/projectiles/guns/energy.dm [hunk 11]
```diff
@@ -132,6 +142,71 @@
...
+			to_chat(usr, SPAN_WARNING("\The [src] is already connected to \the [rigcharger]!"))
+			return
+		if(!ismob(loc))
```
Comment (FluffyGhoster) [inline code/modules/projectiles/guns/energy.dm:165]:
```suggestion

		if(!ismob(loc))
```

File: code/modules/projectiles/guns/energy.dm [hunk 12]
```diff
@@ -132,6 +142,71 @@
...
+			to_chat(usr, SPAN_WARNING("\The [back_charge] must be worn on the back before a weapon can be connected!"))
+			return
+		to_chat(usr, SPAN_NOTICE("You neatly plug \the [src] into \the [recharger]."))
```
Comment (FluffyGhoster) [inline code/modules/projectiles/guns/energy.dm:168]:
```suggestion

		to_chat(usr, SPAN_NOTICE("You neatly plug \the [src] into \the [recharger]."))
```

File: code/modules/projectiles/guns/energy.dm [hunk 13]
```diff
@@ -132,6 +142,71 @@
...
+	if(rigcharger)
+		to_chat(usr, SPAN_NOTICE("With a snap, \the [src] is disconnected from \the [rigcharger]."))
+		playsound(get_turf(src), 'sound/machines/click.ogg', 30, 0)
+		if(rigcharger.active)
+			rigcharger.deactivate()
+		rigcharger.connected = null
+		rigcharger = null
+		self_recharge = initial(self_recharge)
+		use_external_power = initial(use_external_power)
+	if(backcharger)
```
Comment (FluffyGhoster) [inline code/modules/projectiles/guns/energy.dm:185]:
```suggestion

	if(backcharger)
```

File: code/modules/projectiles/guns/energy.dm [hunk 14]
```diff
@@ -132,6 +142,71 @@
...
+		to_chat(usr, SPAN_NOTICE("With a snap, \the [src] is disconnected from \the [backcharger]."))
+		playsound(get_turf(src), 'sound/machines/click.ogg', 30, 0)
+		backcharger.connected = null
+		backcharger = null
+		self_recharge = initial(self_recharge)
+		use_external_power = initial(use_external_power)
+
+/obj/item/gun/energy/MouseDrop(atom/over)
+	. = ..()
+	if(istype(over, /obj/item/rig))
```
Comment (FluffyGhoster) [inline code/modules/projectiles/guns/energy.dm:195]:
```suggestion

	if(istype(over, /obj/item/rig))
```

File: code/modules/projectiles/guns/energy.dm [hunk 15]
```diff
@@ -132,6 +142,71 @@
...
+		var/obj/item/rig/rig = over
+		var/obj/item/rig_module/recharger/recharger = locate() in rig.installed_modules
```
Comment (FluffyGhoster) [inline code/modules/projectiles/guns/energy.dm:197]:
```suggestion

		var/obj/item/rig_module/recharger/recharger = locate() in rig.installed_modules
```

File: code/modules/projectiles/guns/energy.dm [hunk 16]
```diff
@@ -132,6 +142,71 @@
...
+		if(recharger)
+			connect(recharger)
+	if(istype(over, /obj/item/recharger_backpack))
```
Comment (FluffyGhoster) [inline code/modules/projectiles/guns/energy.dm:200]:
```suggestion

	if(istype(over, /obj/item/recharger_backpack))
```

File: code/modules/projectiles/guns/energy.dm [hunk 17]
```diff
@@ -132,6 +142,71 @@
...
+		var/obj/item/recharger_backpack = over
+		connect(recharger_backpack)
+
+
+/obj/item/gun/energy/dropped(mob/living/user)
+	. = ..()
+	if(rigcharger || backcharger)
```
Comment (FluffyGhoster) [inline code/modules/projectiles/guns/energy.dm:207]:
```suggestion

	if(rigcharger || backcharger)
```

File: code/game/objects/items/recharger_backpack.dm [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+	var/obj/item/cell/powersupply
+	var/obj/item/gun/energy/connected
+
+/obj/item/recharger_backpack/Initialize()
+	. = ..()
+	update_icon()
+
+/obj/item/recharger_backpack/examine(mob/user)
+	. = ..()
+	if(powersupply)
```
Comment (FluffyGhoster) [inline code/game/objects/items/recharger_backpack.dm:19]:
```suggestion

	if(powersupply)
```

File: code/game/objects/items/recharger_backpack.dm [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+	item_state = "recharger_backpack"
+	contained_sprite = TRUE
+	w_class = ITEMSIZE_LARGE
+	slot_flags = SLOT_BACK
+	var/obj/item/cell/powersupply
+	var/obj/item/gun/energy/connected
+
+/obj/item/recharger_backpack/Initialize()
+	. = ..()
+	update_icon()
```
Comment (FluffyGhoster) [inline code/game/objects/items/recharger_backpack.dm:15]:
```suggestion

	//To update the icon based on the power cell charge that we spawn with
	update_icon()
```

File: code/modules/projectiles/guns/energy.dm [hunk 20]
```diff
@@ -32,6 +32,8 @@
 	/// Multiplies by charge cost to determine how much charge should be returned
 	var/recharge_multiplier = 1
 	var/charge_tick = 0
+	var/obj/item/rig_module/recharger/rigcharger
+	var/obj/item/recharger_backpack/backcharger
```
Comment (FluffyGhoster) [inline code/modules/projectiles/guns/energy.dm:36]: DMdoc them

File: code/modules/projectiles/guns/energy.dm [hunk 21]
```diff
@@ -119,6 +123,12 @@
 	if(isrobot(src.loc.loc)) // for things inside a robot's module
 		var/mob/living/silicon/robot/R = src.loc.loc
 		return R.cell
+	if(backcharger)
```
Comment (FluffyGhoster) [inline code/modules/projectiles/guns/energy.dm:126]:
```suggestion

	if(backcharger)
```

File: code/modules/projectiles/guns/energy.dm [hunk 22]
```diff
@@ -119,6 +123,12 @@
...
+		return backcharger.powersupply
+	if(rigcharger && rigcharger.holder)
```
Comment (FluffyGhoster) [inline code/modules/projectiles/guns/energy.dm:128]:
```suggestion

	if(rigcharger && rigcharger.holder)
```

File: code/game/objects/items/recharger_backpack.dm [hunk 23]
```diff
@@ -0,0 +1,98 @@
...
+	name = "weapon recharger backpack"
+	desc = "Developed by Hephaestus Industries, the HI-92 Recharge Platform is designed for long-term field recharge of energy weapons. Due to legal concerns, the design is rarely seen outside of military hands."
+	icon = 'icons/obj/recharger_backpack.dmi'
+	icon_state = "recharger_backpack"
+	item_state = "recharger_backpack"
+	contained_sprite = TRUE
+	w_class = ITEMSIZE_LARGE
+	slot_flags = SLOT_BACK
+	var/obj/item/cell/powersupply
+	var/obj/item/gun/energy/connected
```
Comment (FluffyGhoster) [inline code/game/objects/items/recharger_backpack.dm:11]: DMdoc them

File: code/game/objects/items/recharger_backpack.dm [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+		connect(I)
+		return
+
+	return ..()
+
+/obj/item/recharger_backpack/proc/connect(obj/item/gun/energy/newgun)
+	if(connected)
+		to_chat(usr, SPAN_WARNING("\The [src] already has an energy weapon connected!"))
+		return
+	connected = newgun
```
Comment (FluffyGhoster) [inline code/game/objects/items/recharger_backpack.dm:47]:
```suggestion

	connected = newgun
```

File: code/game/objects/items/recharger_backpack.dm [hunk 25]
```diff
@@ -0,0 +1,101 @@
...
+		user.put_in_hands(powersupply)
+		powersupply = null
+		update_icon()
+
+	else if(istype(I, /obj/item/gun/energy))
+		connect(I)
+
+	else
+		. = ..()
+
```
Comment (NonQueueingMatt) [inline code/game/objects/items/recharger_backpack.dm:45]:
```suggestion
```

File: code/modules/projectiles/guns/energy.dm [hunk 26]
```diff
@@ -33,6 +33,10 @@
 	var/recharge_multiplier = 1
 	var/charge_tick = 0
 
+	/// External power sources for the gun. Checked by get_external_power_supply()
+	var/obj/item/rig_module/recharger/rigcharger
+	var/obj/item/recharger_backpack/backcharger
```
Comment (NonQueueingMatt) [inline code/modules/projectiles/guns/energy.dm:38]: You don't need these two to be separate - either unify these variables as a single item then use get_cell() to get the cell from it, or make the rigcharger just be a backpack charger but in the rig.

**Review by NonQueueingMatt — CHANGES_REQUESTED** [review]

**Review by NonQueueingMatt — APPROVED** [review]

**Review by Alberyk — APPROVED** [review]

**Pull request merged**